### PR TITLE
Fix passing `Accelerator` objects to the accelerator flag with `devices=x` & resolve training type plugin

### LINKED ISSUE #10592: Support passing `Accelerator` objects to the `accelerator` flag with `devices=x`
### Environment

- PyTorch Lightning Version (e.g., 1.3.0):
- PyTorch Version (e.g., 1.8)
- Python version:
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### LINKED ISSUE #10775: Resolve training type plugin when passed with Accelerator
### To Reproduce

```python
trainer = Trainer(accelerator=GPUAccelerator(precision_plugin=PrecisionPlugin(), training_type_plugin=DDPPlugin()), gpus=4)
```

At the moment, when training type plugin is passed with Accelerators, attributes such as `parallel_devices`, `cluster_environment` and `sync_batchnorm` are not set to the training plugin and leads to errors.

### Expected behavior

```python
trainer = Trainer(accelerator=GPUAccelerator(precision_plugin=PrecisionPlugin(), training_type_plugin=DDPPlugin()), gpus=4)

# should be equivalent to

training_type_plugin.parallel_devices == [torch.device("cuda", i) for i in self.parallel_device_ids]
training_type_plugin.cluster_environment == LightningEnvironment()
training_type_plugin.sync_batchnorm == False
```

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #10592
Fixes #10775 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @tchaton @akihironitta

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -250,16 +254,24 @@ def _warn_if_devices_flag_ignored(self) -> None:
         if self.devices is None:
             return
         devices_warning = f"The flag `devices={self.devices}` will be ignored, as you have set"
-        if self.distributed_backend in ("auto", _AcceleratorType.TPU):
+        if self.distributed_backend in ("auto", _AcceleratorType.TPU) or isinstance(
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:257]: what if we call `select_accelerator_type` first and then call `_warn_if_devices_flag_ignored`, then we can just use `self.use_ipu/cpu/gpu/tpu` here? will it have any side-effects?

Comment (kaushikb11) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:257]: yeah, we can't as we call `_map_devices_to_accelerator` in `select_accelerator_type`, to assign gpus, ipus, tpu_cores when devices are passed to Trainer.

File: tests/accelerators/test_accelerator_connector.py [hunk 2]
```diff
@@ -1064,3 +1064,79 @@ def test_devices_auto_choice_gpu(is_gpu_available_mock, device_count_mock):
...
+    assert isinstance(trainer.training_type_plugin, IPUPlugin)
+    assert isinstance(trainer.training_type_plugin.cluster_environment, LightningEnvironment)
+    assert trainer.training_type_plugin.parallel_devices == list(range(4))
+
+
+@mock.patch("pytorch_lightning.trainer.connectors.accelerator_connector._TPU_AVAILABLE", return_value=True)
+@mock.patch("pytorch_lightning.utilities.device_parser._TPU_AVAILABLE", return_value=True)
+@mock.patch("pytorch_lightning.plugins.training_type.tpu_spawn._TPU_AVAILABLE", return_value=True)
+def test_training_plugin_attrs_set_when_passed_with_tpu_accelerator(
+    acc_connector_tpu_mock, device_parser_tpu_mock, training_type_tpu_available_mock
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply tests/accelerators/test_accelerator_connector.py]: Updated!

File: tests/accelerators/test_accelerator_connector.py [hunk 3]
```diff
@@ -1064,3 +1064,79 @@ def test_devices_auto_choice_gpu(is_gpu_available_mock, device_count_mock):
     trainer = Trainer(accelerator="auto", devices="auto")
     assert trainer.devices == 2
     assert trainer.gpus == 2
+
+
+def test_training_plugin_attrs_set_when_passed_with_cpu_accelerator():
+    trainer = Trainer(accelerator=CPUAccelerator(PrecisionPlugin(), DDPPlugin()), num_processes=3)
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply tests/accelerators/test_accelerator_connector.py:1070]: Yes yes, let's update all accelerator calls in tests post the refactor

File: CHANGELOG.md [hunk 4]
```diff
@@ -202,6 +202,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `_compare_version` for python packages ([#10762](https://github.com/PyTorchLightning/pytorch-lightning/pull/10762))
 
 
+- Fixed passing `Accelerator` objects to the accelerator flag with `devices=x` ([#10773](https://github.com/PyTorchLightning/pytorch-lightning/pull/10773))
```
Comment (awaelchli) [inline CHANGELOG.md:205]: I agree with this fix

File: CHANGELOG.md [hunk 5]
```diff
@@ -202,6 +202,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
+
+
+- Resolve training type plugin when passed within `Accelerator` ([#10773](https://github.com/PyTorchLightning/pytorch-lightning/pull/10773))
```
Comment (awaelchli) [inline CHANGELOG.md:208]: I am confused by this fix. IMO it is not worth doing that considering the refactors that happen right now for 1.6. I could however see this being included in 1.5.x

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md (hunk 5 above)
Comment (kaushikb11) [reply CHANGELOG.md:208]: @awaelchli But it would be required for 1.5.x, right?

Comment (awaelchli) [reply CHANGELOG.md:208]: Yes sure, it can be merged to master first, then be included in 1.5.x, but then tests that you add here become mostly outdated immediately after. That's all I'm saying.

File: tests/accelerators/test_accelerator_connector.py [hunk 6]
```diff
@@ -1064,3 +1064,75 @@ def test_devices_auto_choice_gpu(is_gpu_available_mock, device_count_mock):
...
+    assert isinstance(trainer.training_type_plugin, DDPPlugin)
+    assert isinstance(trainer.training_type_plugin.cluster_environment, LightningEnvironment)
+    assert trainer.training_type_plugin.parallel_devices == [
+        torch.device(type="cpu"),
+        torch.device(type="cpu"),
+        torch.device(type="cpu"),
+    ]
+    assert not trainer.training_type_plugin.sync_batchnorm
+
+    trainer = Trainer(accelerator=CPUAccelerator(PrecisionPlugin(), DDPPlugin()), devices=2, sync_batchnorm=True)
```
Comment (tchaton) [inline tests/accelerators/test_accelerator_connector.py:1080]:
```suggestion
    trainer = Trainer(accelerator=CPUAccelerator(PrecisionPlugin(), DDPPlugin()), devices=2)
```

**Review by tchaton — APPROVED** [review]
LGTM !

File: CHANGELOG.md (hunk 5 above)
Comment (carmocca) [reply CHANGELOG.md:208]: In that case, let's just change the target branch to the bugfix release branch

File: tests/accelerators/test_accelerator_connector.py (hunk 6 above)
Comment (kaushikb11) [reply tests/accelerators/test_accelerator_connector.py:1080]: Why remove it?

**Review by Borda — APPROVED** [review]

**Comment by kaushikb11** [comment]
Closing it. As the issue has been resolved after the AcceleratorConnector refactor
